Indexing/SubstitutionTree.hpp

```diff
@@ -108,200 +108,193 @@ using namespace Kernel;
+  }
+  SubstitutionTree& operator=(SubstitutionTree && other) { swap(*this,other); return *this; }
+  SubstitutionTree(SubstitutionTree&& other) : SubstitutionTree(0) { swap(*this, other); }
 
-      LeafData(Clause* cls, Literal* literal)
-      : clause(cls), literal(literal), term(TermList::empty()), sort(TermList::empty()), extraTerm(TermList::empty()) {  }
-      inline
-      bool operator==(const LeafData& o)
-      { return clause==o.clause && literal==o.literal && term==o.term; }
+  ~SubstitutionTree()
+  {
+    ASS_EQ(_iterCnt,0);
+    delete _root;
+  }
 
-      Clause* clause;
-      Literal* literal;
-      TermList term;
-      TermList sort;
-      // In some higher-order use cases, we want to store a different term 
-      // in the leaf to the indexed term. extraTerm is used for this purpose.
-      // In all other situations it is empty
-      TermList extraTerm;
-
-      friend std::ostream& operator<<(std::ostream& out, LeafData const& self)
-      { return out << outputPtr(self.clause); }
-    };
-    typedef VirtualIterator<LeafData*> LDIterator;
+#define VERBOSE_OUTPUT_OPERATORS 0
+  friend std::ostream& operator<<(std::ostream& out, SubstitutionTree const& self)
+  {
+#if VERBOSE_OUTPUT_OPERATORS
+    out << "{ nextVar: S" << self._nextVar << ", root: (";
+#endif // VERBOSE_OUTPUT_OPERATORS
+    if (self._root) {
+      out << *self._root;
+    } else {
+      out << "<empty tree>";
+    }
+#if VERBOSE_OUTPUT_OPERATORS
+    out << ") }";
+#endif // VERBOSE_OUTPUT_OPERATORS
+    return out;
+  }
+#undef VERBOSE_OUTPUT_OPERATORS
+  friend std::ostream& operator<<(std::ostream& out, OutputMultiline<SubstitutionTree> const& self)
+  {
+    if (self.self._root) {
+      self.self._root->output(out, true, /* indent */ 0);
+    } else {
+      out << "<empty tree>";
+    }
+    return out;
+  }
 
-    template<class Unifier>
-    struct QueryResult {
-      LeafData const* data; 
-      Unifier unif;
 
-      QueryResult(LeafData const* ld, Unifier unif) : data(ld), unif(std::move(unif)) {}
-    };
-    template<class Unifier>
-    static QueryResult<Unifier>  queryResult(LeafData const* ld, Unifier unif) 
-    { return QueryResult<Unifier>(ld, std::move(unif)); }
+  template<class T>
+  friend std::ostream& operator<<(std::ostream& out, SubstitutionTree<T> const& self);
+  template<class T>
+  friend std::ostream& operator<<(std::ostream& out, OutputMultiline<SubstitutionTree<T>> const& self);
+  typedef VirtualIterator<LeafData*> LDIterator;
 
-    template<class I, class TermOrLit, class... Args> 
-    inline auto iterator(TermOrLit query, bool retrieveSubstitutions, bool reversed, Args... args)
-    { return ifElseIter(_root == nullptr, 
-        [&](){ return EmptyIter<ELEMENT_TYPE(I)>{}; },
-        [&](){ return iterPointer(std::make_unique<I>(this, _root, query, retrieveSubstitutions, reversed, std::move(args)...)); }); }
+  template<class Unifier>
+  struct QueryResult {
+    LeafData const* data; 
```

Comment:
If we're redesigning the query interface anyway, is it feasible/worth making this `const LeafData &data`?

Reply:
No objection, as long as it's considered rather than oversight.